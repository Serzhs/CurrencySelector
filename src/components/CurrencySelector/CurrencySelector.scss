$itemHeight: 35px;
$c-backGround: #e5e5e5;
$c-backGroundHovered: #e5e5e5;
$c-active: #ffffff;
$c-darkShade: #333333;
$c-border: #a3a3a3;
$c-selectedShade: #ff0000;
$c-pressable-text: #134371;

.container {
  border: 1px solid $c-border;
  padding: 1rem;
  width: 36rem;
  border-radius: 5px;
  box-shadow: 3px 3px 5px 0 rgba(0,0,0,0.2);
}

.selectedItemsWrapper {
  display: flex;
  flex-wrap: wrap;
  flex-basis: 100%;
  max-width: 100%;
  margin-bottom: 1rem;
  min-height: 2rem;
  justify-content: flex-start;
}

.selectedItem {
  background-color: $c-backGround;
  min-height: $itemHeight;
  width: 98%;
  position: relative;
  text-transform: uppercase;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;
}

.selectedItemClose {
  position: absolute;
  right: -0.5rem;
  top: -0.5rem;
  background-color: $c-darkShade;
  color: $c-active;
  width: 1.625rem;
  height: 1.625rem;
  border-radius: 50%;
  border-style: solid;
  border-width: 2px;
  border-color: $c-active;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  line-height: 1;
  text-transform: lowercase;
  transition: color 0.2s, background-color 0.2s, border-color 0.2s;

  &:hover {
    color: $c-darkShade;
    border-color: $c-darkShade;
    background-color: $c-active;
  }
}

.itemsWrapper {
  display: flex;
  flex-basis: 100%;
  max-width: 100%;
  flex-wrap: wrap;
}

.itemWrapper {
  max-width: 33.333%;
  flex-basis: 33.333%;
  padding: 0.5rem;
}

.item {
  border: 1px solid $c-border;
  background-color: $c-backGround;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  padding: 0.5rem;
  text-transform: uppercase;
  line-height: 1;
  color: $c-pressable-text;
  font-size: 1.125rem;
  cursor: pointer;
  transition: background-color 0.2s;

  &:hover {
    background-color: $c-backGroundHovered;
  }

  &.itemSelected {
    background-color: $c-active;

    .itemCheckbox:before {
      content: 'X';
    }

  }
}

.itemCheckbox {
  position: relative;
  width: 1rem;
  height: 1rem;
  border: 1px solid $c-border;
  border-radius: 3px;
  margin-right: 0.3125rem;
  color: $c-selectedShade;
  font-weight: bold;
  font-size: 0.875rem;
  display: flex;
  align-items: center;
  justify-content: center;
  line-height: 1;
}
