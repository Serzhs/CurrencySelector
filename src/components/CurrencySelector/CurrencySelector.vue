<template>
  <div :class="styles.container">
    <div :class="styles.selectedItemsWrapper">
      <div
        v-for="{ name } in selectedCurrencies"
        :key="`selected-${name}`"
        :class="styles.itemWrapper"
      >
        <div :class="styles.selectedItem">
          <span testID="selectedCurrencies">
            {{ name }}
          </span>
          <span
            @click="removeSelectedCurrency(name)"
            :class="styles.selectedItemClose"
            testID="selectedCurrenciesRemove"
          >
            x
          </span>
        </div>
      </div>
    </div>
    <div :class="styles.itemsWrapper">
      <label
        v-for="({ name, selected }, i) in currencies"
        :key="name"
        :for="name"
        :class="styles.itemWrapper"
        testID="currencies"
      >
        <span :class="[styles.item, selected && styles.itemSelected]">
          <input
            type="checkbox"
            :id="name"
            v-model="currencies[i].selected"
            hidden
            :class="styles.itemCheckboxInput"
          />
          <span :class="styles.itemCheckbox" />
          {{ name }}
        </span>
      </label>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import styles from "./CurrencySelector.scss";

type CurrencyNames =
  | "EUR"
  | "PLN"
  | "GEL"
  | "DKK"
  | "CZK"
  | "GBP"
  | "SEK"
  | "USD"
  | "RUB";

type Currency = {
  name: CurrencyNames;
  selected: boolean;
};

declare interface BaseComponentData {
  styles: { [className: string]: string };
  currencies: Currency[];
}

export default Vue.extend({
  name: "CurrencySelector",
  data(): BaseComponentData {
    return {
      styles: styles,
      currencies: [
        {
          name: "EUR",
          selected: false
        },
        {
          name: "PLN",
          selected: false
        },
        {
          name: "GEL",
          selected: false
        },
        {
          name: "DKK",
          selected: false
        },
        {
          name: "CZK",
          selected: false
        },
        {
          name: "GBP",
          selected: false
        },
        {
          name: "SEK",
          selected: false
        },
        {
          name: "USD",
          selected: false
        },
        {
          name: "RUB",
          selected: false
        }
      ]
    };
  },
  computed: {
    selectedCurrencies(): Currency[] {
      return this.currencies.filter(({ selected }) => selected);
    }
  },
  methods: {
    removeSelectedCurrency(currencyName: CurrencyNames) {
      const currencyToUnselectIndex = this.currencies.findIndex(
        ({ name }) => name === currencyName
      );

      const newCurrencyValue: Currency = {
        name: currencyName,
        selected: false
      };

      this.$set(this.currencies, currencyToUnselectIndex, newCurrencyValue);
    }
  }
});
</script>
