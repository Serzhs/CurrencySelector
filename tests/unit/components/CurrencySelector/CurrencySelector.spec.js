import { shallowMount } from "@vue/test-utils";
import CurrencySelector from "@/components/CurrencySelector/CurrencySelector";

describe("CurrencySelector.vue", () => {
  it("Check if exists", () => {
    const component = shallowMount(CurrencySelector);
    expect(component.vm).toBeTruthy();
  });
  it("Matches snapshot", () => {
    const component = shallowMount(CurrencySelector);
    expect(component.element).toMatchSnapshot();
  });
});
