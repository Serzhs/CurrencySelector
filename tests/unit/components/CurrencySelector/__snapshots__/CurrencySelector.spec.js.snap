// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelector.vue Matches snapshot 1`] = `
<div>
  <div />
   
  <div>
    <label
      for="EUR"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="EUR"
          type="checkbox"
        />
         
        <span />
        
        EUR
      
      </span>
    </label>
    <label
      for="PLN"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="PLN"
          type="checkbox"
        />
         
        <span />
        
        PLN
      
      </span>
    </label>
    <label
      for="GEL"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="GEL"
          type="checkbox"
        />
         
        <span />
        
        GEL
      
      </span>
    </label>
    <label
      for="DKK"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="DKK"
          type="checkbox"
        />
         
        <span />
        
        DKK
      
      </span>
    </label>
    <label
      for="CZK"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="CZK"
          type="checkbox"
        />
         
        <span />
        
        CZK
      
      </span>
    </label>
    <label
      for="GBP"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="GBP"
          type="checkbox"
        />
         
        <span />
        
        GBP
      
      </span>
    </label>
    <label
      for="SEK"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="SEK"
          type="checkbox"
        />
         
        <span />
        
        SEK
      
      </span>
    </label>
    <label
      for="USD"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="USD"
          type="checkbox"
        />
         
        <span />
        
        USD
      
      </span>
    </label>
    <label
      for="RUB"
      testid="currencies"
    >
      <span
        class=""
      >
        <input
          hidden="hidden"
          id="RUB"
          type="checkbox"
        />
         
        <span />
        
        RUB
      
      </span>
    </label>
  </div>
</div>
`;
